server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java

```diff
@@ -1212,6 +1212,109 @@ public Settings onNodeStopped(String nodeName) throws Exception {
+        } else {
+            reasonableOperationsBasedRecoveryProportion
+                = IndexSettings.FILE_BASED_RECOVERY_THRESHOLD_SETTING.get(Settings.EMPTY);
+        }
+        logger.info("--> performing ops-based recoveries up to [{}%] of docs", reasonableOperationsBasedRecoveryProportion * 100.0);
+
+        createIndex(indexName, settings.build());
+        indexRandom(randomBoolean(), false, randomBoolean(), IntStream.range(0, between(0, 100))
+            .mapToObj(n -> client().prepareIndex(indexName, "_doc").setSource("num", n)).collect(toList()));
+        ensureGreen(indexName);
+
+        flush(indexName);
+        // wait for all history to be discarded
+        assertBusy(() -> {
+            for (ShardStats shardStats : client().admin().indices().prepareStats(indexName).get().getShards()) {
+                final long maxSeqNo = shardStats.getSeqNoStats().getMaxSeqNo();
+                assertTrue(shardStats.getRetentionLeaseStats().retentionLeases() + " should discard history up to " + maxSeqNo,
+                    shardStats.getRetentionLeaseStats().retentionLeases().leases().stream().allMatch(
+                        l -> l.retainingSequenceNumber() == maxSeqNo + 1));
+            }
+        });
+        flush(indexName); // ensure that all operations are in the safe commit
+
+        final ShardStats shardStats = client().admin().indices().prepareStats(indexName).get().getShards()[0];
+        final long docCount = shardStats.getStats().docs.getCount();
+        assertThat(shardStats.getStats().docs.getDeleted(), equalTo(0L));
+        assertThat(shardStats.getSeqNoStats().getMaxSeqNo() + 1, equalTo(docCount));
+
+        final ShardId shardId = new ShardId(resolveIndex(indexName), 0);
+        final DiscoveryNodes discoveryNodes = clusterService().state().nodes();
+        final IndexShardRoutingTable indexShardRoutingTable = clusterService().state().routingTable().shardRoutingTable(shardId);
+
+        final ShardRouting replicaShardRouting = indexShardRoutingTable.replicaShards().get(0);
+        assertTrue("should have lease for " + replicaShardRouting,
+            client().admin().indices().prepareStats(indexName).get().getShards()[0].getRetentionLeaseStats()
+                .retentionLeases().contains(ReplicationTracker.getPeerRecoveryRetentionLeaseId(replicaShardRouting)));
+        internalCluster().restartNode(discoveryNodes.get(replicaShardRouting.currentNodeId()).getName(),
+            new InternalTestCluster.RestartCallback() {
+                @Override
+                public Settings onNodeStopped(String nodeName) throws Exception {
+                    assertFalse(client().admin().cluster().prepareHealth()
+                        .setWaitForNodes(Integer.toString(discoveryNodes.getSize() - 1))
+                        .setWaitForEvents(Priority.LANGUID).get().isTimedOut());
+
+                    final int newDocCount = Math.toIntExact(Math.round(Math.ceil(
```

Comment:
I could worry that an inaccuracy from using doubles here and in `ReplicationTracker` could result in this failing rarely. For instance if `docCount * reasonableOperationsBasedRecoveryProportion` is slightly off such that it is just below an integer number rather than just above. I think using a less random percentage (ie. a percentage that does not use the full double precision) could solve this. I might also be seeing ghosts here, so am OK to leave this for now provided this has been run successfully a number of times.

Reply:
I didn't do a full analysis of the rounding errors here, but I did run the following simulation hundreds of times without seeing a failure. The calculated bound is not tight.

```
    public void testNewDocCountCalculation() {
        for (int i = 0; i < 1000000; i++) {
            final int docCount = between(1, 100);
            final double reasonableOperationsBasedRecoveryProportion = randomDoubleBetween(0.05, 0.99, true);
            final int newDocCount = Math.toIntExact(Math.round(Math.ceil(
                (1 + Math.ceil(docCount * reasonableOperationsBasedRecoveryProportion))
                    / (1 - reasonableOperationsBasedRecoveryProportion))));

            assertThat(docCount, lessThan(docCount + newDocCount
                - Math.toIntExact(Math.round(Math.ceil((docCount + newDocCount) * reasonableOperationsBasedRecoveryProportion)))));
        }
    }
```

Reply:
Thanks, I think that is adequate.